### PR TITLE
some tests and readme

### PR DESCRIPTION
**Pull request by sarimabbas**
added tests to test_utils_lib, begun test_gripper, created read me, modified CMakeLists for tests

### CODE REVIEW

**Comment by alecive** [comment]
@sarimabbas nice pic! In general, it is a good practice to do smaller commits that are self-contained (eg one for the readme, one for the unit tests), and to push often, so that I don't have to review everything altogether.

Btw for now it's okay, I'll wait for travis and codacy and then I'll do my review.

**Comment by sarimabbas** [comment]
Oh ok, will do next time!

File: baxter_collaboration_lib/CMakeLists.txt [hunk 1]
```diff
@@ -138,12 +138,19 @@ if(CATKIN_ENABLE_TESTING)
...
 
+  ## Robot util tests
   catkin_add_gtest(test_utils_lib test/test_utils_lib.cpp)
   target_link_libraries(test_utils_lib robot_utils)
 
+  ## Robot interface tests
   add_rostest_gtest(test_robot_interface test/test_robot_interface.test
                                          test/test_robot_interface.cpp)
   target_link_libraries(test_robot_interface  robot_interface ${catkin_LIBRARIES})
+
```
Comment (alecive) [inline baxter_collaboration_lib/CMakeLists.txt:149]: Add comment as you did with `Robot interface tests` and "Robot util test"

File: baxter_collaboration_lib/test/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
+## Testing
+
+#### Using [Catkin Tools](https://catkin-tools.readthedocs.io/en/latest/)
```
Comment (alecive) [inline baxter_collaboration_lib/test/README.md]: Remove one level of heading (also below)

File: baxter_collaboration_lib/test/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+#### Checklist
+* /baxter_collaboration_lib/src/robot_utils
+	- [x] utils.cpp
+	- [ ] baxter_trac_ik.cpp
+	- [ ] ros_thread.cpp
+	- [ ] ros_thread_image.cpp
+	- [ ] ros_thread_obj.cpp
+
+* /baxter_collaboration_lib/src/robot_interface
+	- [x] robot_interface.cpp
```
Comment (alecive) [inline baxter_collaboration_lib/test/README.md]: `robotInterface` tests still need to be done.

File: baxter_collaboration_lib/test/test_gripper.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	string 	limb = "left";
+	bool 	use_robot = false;
+
+	Gripper gr(limb, use_robot);
+
+
+	EXPECT_EQ(false, gr.is_enabled());
+	EXPECT_EQ(false, gr.is_calibrated());
+	EXPECT_EQ(false, gr.is_ready_to_grip());
+	EXPECT_EQ(false, gr.has_error());
```
Comment (alecive) [inline baxter_collaboration_lib/test/test_gripper.cpp]: This is not necessarily your fault, but prefer `EXPECT_FALSE` to `EXPECT_EQ(false, `. If that appeared in `test_robot_interface`, please change that as well.

File: baxter_collaboration_lib/test/test_gripper.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+	EXPECT_EQ(true, gr.is_sucking());
+	EXPECT_EQ(true, gr.is_gripping());
```
Comment (alecive) [inline baxter_collaboration_lib/test/test_gripper.cpp]: Simliarly, use `EXPECT_TRUE` here.

File: baxter_collaboration_lib/test/test_gripper.cpp [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+	EXPECT_EQ("left", gr.getGripperLimb());
+
+	//EXPECT_EQ(true, gr.gripObject());
+	//EXPECT_EQ(true, gr.releaseObject());
```
Comment (alecive) [inline baxter_collaboration_lib/test/test_gripper.cpp]: Delete these comments. These should be tested in the simulator test.

File: baxter_collaboration_lib/test/test_utils_lib.cpp [hunk 7]
```diff
@@ -89,6 +89,62 @@ TEST(UtilsLib, State)
...
+    testDoubleVector.push_back(10.123);
+    testDoubleVector.push_back(20.123);
+    EXPECT_EQ("[10.123, 20.123]", toString(testDoubleVector));
+
+    geometry_msgs::Pose testPose;
+    testPose.position.x = 0; testPose.position.y = 5; testPose.position.z = 10;
+    testPose.orientation.x = 15; testPose.orientation.y = 20; testPose.orientation.z = 25; 
+    testPose.orientation.w = 30;
+    EXPECT_EQ("{position:{x: 0, y: 5, z: 10}, orientation:{x: 15, y: 20, z: 25, w: 30}}", toString(testPose));
+}
```
Comment (alecive) [inline baxter_collaboration_lib/test/test_utils_lib.cpp]: Good job here! But try also to think about edge cases, eg negative numbers, zeros, things that might break the code.

**Review by alecive — CHANGES_REQUESTED** [review]

**Comment by sarimabbas** [comment]
Have reviewed everything except the edge cases for test_utils_lib, will think about them in detail today!

File: baxter_collaboration_lib/test/test_utils_lib.cpp [hunk 8]
```diff
@@ -89,6 +89,62 @@ TEST(UtilsLib, State)
 
 }
 
+// sarim's tests
```
Comment (alecive) [inline baxter_collaboration_lib/test/test_utils_lib.cpp]: Remove this comment, it's not useful :)

File: baxter_collaboration_lib/test/README.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+Troubleshooting:
+* If presented with a `roslaunch` error, navigate to `~/.bashrc` and make sure the following is appended to the file:
+	```
+	source /opt/ros/indigo/setup.bash
+	source ~/ros_ws/devel/setup.bash
+	```
+* If seeing unchanging results, `catkin build` may be required.
+
+### Checklist
```
Comment (alecive) [inline baxter_collaboration_lib/test/README.md]: I would probably move the checklist out of the `README`. Open an issue and place it there instead.

**Review by alecive — CHANGES_REQUESTED** [review]

File: baxter_collaboration_lib/include/robot_utils/utils.h [hunk 10]
```diff
@@ -107,6 +107,16 @@ void setPosition(geometry_msgs::Pose& pose, float x, float y, float z);
...
 
+/**
+ * replaces all instances of a substring with specified substring
+ * @param target      string containing the substring
+ * @param replaceThis substring to be replaced
+ * @param replaceWith substring to replace with
+ *
+ * return N/A
+ */
+void subStringReplace(std::string& target, std::string replaceThis, std::string replaceWith);
```
Comment (alecive) [inline baxter_collaboration_lib/include/robot_utils/utils.h]: What is this function for?

File: baxter_collaboration_lib/test/README.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+### Using [Catkin Tools](https://catkin-tools.readthedocs.io/en/latest/)
+
+1. Navigate to your workspace: `cd ros_ws`
+2. `REPO_NAME=baxter_collaboration`
+3. Compile and test: `catkin build "$REPO_NAME"_lib && catkin build "$REPO_NAME"_lib --catkin-make-args run_tests`
+4. Check the results of above: `catkin_test_results build/"$REPO_NAME"_lib || (cat build/"$REPO_NAME"_lib/test_results/"$REPO_NAME"_lib/*.xml ; false)`
+
+Troubleshooting:
+* If build fails, make sure prerequisites are installed: [`aruco_ros`](https://github.com/ScazLab/aruco_ros), `nlopt`, [`trac_ik`](https://bitbucket.org/alecive/trac_ik), [`baxter_common`](http://sdk.rethinkrobotics.com/wiki/Workstation_Setup#adeedda5360641914fe9c5d681c30026) and [`moveit`](http://moveit.ros.org/install/).
```
Comment (alecive) [inline baxter_collaboration_lib/test/README.md]: Why`moveit`?

**Review by alecive — CHANGES_REQUESTED** [review]

File: baxter_collaboration_lib/include/robot_utils/utils.h (hunk 10 above)
Comment (alecive) [reply baxter_collaboration_lib/include/robot_utils/utils.h]: I'm not saying that you should not have done it, but please justify :)

Comment (sarimabbas) [reply baxter_collaboration_lib/include/robot_utils/utils.h]: Sure. I wanted to preserve the code that was already present that used stringstream. But the str() method with stringstream would not convert something like -0.0 to 0 (instead it would give -0). So for just the toString(double) situation, the fix was easy enough, but for the toString(doubleVector) and toString(Pose) situations, there was the possibility of multiple -0s being present in the final string. So I thought I'd create a new function in utils that I could use for both situations.

Within the toString functions, a vector like [-0.0, 1, 2, 3, -0] now becomes [0, 1, 2, 3, 0]. Same for a Pose. I tried to think of potential dangers of such a function, but in my use, I have only targeted "-0," and "-0}" substrings for replacement.

File: baxter_collaboration_lib/test/README.md (hunk 11 above)
Comment (sarimabbas) [reply baxter_collaboration_lib/test/README.md]: When Andrew and I were trying to get him set up, he was unable to build baxter_collaboration without moveit. On the other hand, his trac_ik package was not building, so perhaps if his trac_ik builds, moveit may not be required. So I can defer this till tomorrow?

Comment (alecive) [reply baxter_collaboration_lib/test/README.md]: Yes. Moveit is not needed in our software, let's take a look at that together.

File: baxter_collaboration_lib/include/robot_utils/utils.h (hunk 10 above)
Comment (alecive) [reply baxter_collaboration_lib/include/robot_utils/utils.h]: What if you change `toString(vector)` into a for loop that cycles through the elements and does individual `toString(double)`? Overall, this is a very minor function, whose behavior is not critical since it is only used for printouts. I would try to move on with this as fast as possible and start testing more important things.

Comment (sarimabbas) [reply baxter_collaboration_lib/include/robot_utils/utils.h]: You're right, this was a cleaner approach which I have implemented :)

**Comment by alecive** [comment]
Ok feel free to merge, delete the branch and work on a new thing in a new branch.

**Comment by sarimabbas** [comment]
Great! How would i do the merge? The button is greyed out on my end.

**Review by alecive — APPROVED** [review]

**Comment by alecive** [comment]
Ops sorry now it should be working.

**Comment by sarimabbas** [comment]
screenshot from 2017-05-23 12 57 40

Sorry but even after your approval it is greyed out
